codec/src/test/java/io/netty/handler/codec/compression/BrotliEncoderTest.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.aayushatharva.brotli4j.decoder.Decoder;
+import com.aayushatharva.brotli4j.decoder.DecoderJNI;
+import com.aayushatharva.brotli4j.decoder.DirectDecompress;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.embedded.EmbeddedChannel;
+
+import java.io.IOException;
+
+public class BrotliEncoderTest extends AbstractEncoderTest {
+
+    static {
+        try {
+            Brotli.ensureAvailability();
+        } catch (Throwable throwable) {
+            throw new ExceptionInInitializerError(throwable);
+        }
+    }
+
+    @Override
+    public EmbeddedChannel createChannel() {
+        return new EmbeddedChannel(new BrotliEncoder());
+    }
+
+    protected ByteBuf decompress(ByteBuf compressed, int originalLength) throws Exception {
+        byte[] compressedArray = new byte[compressed.readableBytes()];
+        compressed.readBytes(compressedArray);
+        compressed.release();
+
+        DirectDecompress decompress = Decoder.decompress(compressedArray);
+        if (decompress.getResultStatus() == DecoderJNI.Status.ERROR) {
+            throw new DecompressionException("Brotli stream corrupted");
+        }
+
+        byte[] decompressed = decompress.getDecompressedData();
+        return Unpooled.wrappedBuffer(decompressed);
+    }
+
+    @Override
+    public void testCompressionOfBatchedFlowOfData(ByteBuf data) throws Exception {
+        // NO-OP
```

Comment:
IIUC you had to skip this test because `Decoder.decompress(array)` does not expect multiple compressed blocks inside the array. It requires an `array` to be a single compressed block.

To support that use-case, you can override `protected ByteBuf readDecompressed(final int dataLength)` method from `AbstractCompressionTest`. Instead of composing compressed `ByteBuf`s you can do decompression for each `msg` you read from the channel and compose decompressed messages:

```patch
Index: codec/src/test/java/io/netty/handler/codec/compression/BrotliEncoderTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/codec/src/test/java/io/netty/handler/codec/compression/BrotliEncoderTest.java b/codec/src/test/java/io/netty/handler/codec/compression/BrotliEncoderTest.java
--- a/codec/src/test/java/io/netty/handler/codec/compression/BrotliEncoderTest.java	(revision 395c2852903c5df6f71b09cfbac6c14716d7cd71)
+++ b/codec/src/test/java/io/netty/handler/codec/compression/BrotliEncoderTest.java	(date 1622739587115)
@@ -23,8 +23,6 @@
 import io.netty.buffer.Unpooled;
 import io.netty.channel.embedded.EmbeddedChannel;
 
-import java.io.IOException;
-
 public class BrotliEncoderTest extends AbstractEncoderTest {
 
     static {
@@ -40,6 +38,7 @@
         return new EmbeddedChannel(new BrotliEncoder());
     }
 
+    @Override
     protected ByteBuf decompress(ByteBuf compressed, int originalLength) throws Exception {
         byte[] compressedArray = new byte[compressed.readableBytes()];
         compressed.readBytes(compressedArray);
@@ -55,7 +54,16 @@
     }
 
     @Override
-    public void testCompressionOfBatchedFlowOfData(ByteBuf data) throws Exception {
-        // NO-OP
+    protected ByteBuf readDecompressed(final int dataLength) throws Exception {
+        CompositeByteBuf decompressed = Unpooled.compositeBuffer();
+        ByteBuf msg;
+        while ((msg = channel.readOutbound()) != null) {
+            if (msg.isReadable()) {
+                decompressed.addComponent(true, decompress(msg, -1));
+            } else {
+                msg.release();
+            }
+        }
+        return decompressed;
     }
 }
```

Reply:
Thanks a lot!